modules/battery_manager/php/testcontroller.class.inc

```diff
@@ -0,0 +1,251 @@
+ * @author   Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+class TestController implements RequestHandlerInterface
+{
+
+    /**
+     * Returns true if user has access to this endpoint.
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return true;
+    }
+
+    /**
+     * This function passes the request to the handler. This is necessary since
+     * the Endpoint bypass the Module class.
+     *
+     * XXX: This function should be extracted to a parent class.
+     *
+     * @param ServerRequestInterface  $request The PSR7 request.
+     * @param RequestHandlerInterface $handler The request handler.
+     *
+     * @return ResponseInterface The outgoing PSR7 response.
+     */
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        return $handler->handle($request);
+    }
+
+    /**
+     * This acts as an Ajax enpoint that handles all action requests from the
+     * Battery Manager Module.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $db     = \Database::singleton();
+        $user   = $request->getAttribute('user');
+        $method = $request->getMethod();
+        $test   = json_decode($request->getBody()->getContents(), true);
+
+        switch($method) {
+        case 'GET':
+            return $this->getInstances($user);
+        case 'POST':
+            return $this->addInstance($db, $user, $test);
+        case 'PUT':
+            return $this->updateInstance($db, $test);
+        }
+    }
+
+    /**
+     * Gets the data source for battery manager tests.
+     *
+     * @param \User $user User Instance
+     *
+     * @return \LORIS\Data\Provisioner
+     */
+    function getDataProvisioner(\User $user) : \LORIS\Data\Provisioner
+    {
+        $provisioner = new TestProvisioner();
+
+        if ($user->hasPermission('access_all_profiles') == false) {
+            $provisioner = $provisioner->filter(
+                new \LORIS\Data\Filters\UserSiteMatch()
+            );
+        }
+        return $provisioner;
+    }
+
+    /**
+     * Updates the status of the given test from Active to Deactive or
+     * viceversa.
+     *
+     * @param \Database $db   Database Instance
+     * @param array     $data Data to be saved
+     *
+     * @return \LORIS\Http\Response response
+     */
+    function updateInstance(\Database $db, array $data)
+    {
+        $test = $this->toSQL($data);
+        try {
+            $db->update('test_battery', $test, array('ID' => $test['ID']));
```

Comment:
If the provided $data is missing some properties, will it set the values to null? I am asking because it seems that `updateInstance` is called to handle a PUT request and PUT requests expect a new full representation of the resource (test_battery). Otherwise, PATCH should be used to update some fields.

Reply:
Good question, I will look into this.

Reply:
After looking at this again, it seems like the full representation of the `test` should be sent when using this function. However, I don't know that if there's some missing properites, that the values will be set to null. I guess that would depend on how `db->update` works.